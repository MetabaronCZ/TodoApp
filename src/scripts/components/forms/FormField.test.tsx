import React from 'react';

import { fireEvent, render } from '@testing-library/react';
import { describe, expect, it, jest } from '@jest/globals';

import * as Dropdown from 'components/forms/Dropdown';
import { FormField } from 'components/forms/FormField';

import { withMockedProviders } from 'test/component';

describe('components/forms/FormField', () => {
  it('should render correctly', () => {
    const { container } = render(
      withMockedProviders(
        <FormField
          label="Label"
          field={{
            type: 'text',
            value: 'Content',
            onChange: () => null,
          }}
        />,
      ),
    );
    expect(container).toMatchSnapshot();

    const label = container.querySelector('label');
    expect(label).toBeInTheDocument();
    expect(label?.textContent).toEqual('Label:');
  });

  it('should render text input', () => {
    const { container } = render(
      withMockedProviders(
        <FormField
          label="Label"
          field={{
            type: 'text',
            value: 'Content',
            onChange: () => null,
          }}
        />,
      ),
    );
    expect(container).toMatchSnapshot();

    const input = container.querySelector('input');
    expect(input).toBeInTheDocument();
    expect(input?.value).toEqual('Content');
  });

  it('should render textarea', () => {
    const { container } = render(
      withMockedProviders(
        <FormField
          label="Label"
          field={{
            type: 'text',
            value: 'Content',
            textarea: true,
            onChange: () => null,
          }}
        />,
      ),
    );
    expect(container).toMatchSnapshot();

    const textarea = container.querySelector('textarea');
    expect(textarea).toBeInTheDocument();
    expect(textarea?.value).toEqual('Content');
  });

  it('should render dropdown', () => {
    const mockedDropdown = jest.spyOn(Dropdown, 'Dropdown');
    expect(mockedDropdown).toBeCalledTimes(0);

    const { container } = render(
      withMockedProviders(
        <FormField
          label="Label"
          field={{
            type: 'dropdown',
            value: '1',
            options: [{ id: '1', title: 'Item 1', value: '1' }],
            onChange: () => null,
          }}
        />,
      ),
    );
    expect(container).toMatchSnapshot();

    expect(mockedDropdown).toBeCalledTimes(1);
    const props = mockedDropdown.mock.calls[0][0];
    expect(props.items.length).toEqual(1);
  });

  it('should handle text input change', () => {
    const onChange = jest.fn();

    const { container } = render(
      withMockedProviders(
        <FormField
          label="Label"
          field={{
            type: 'text',
            value: 'Content',
            onChange,
          }}
        />,
      ),
    );
    expect(onChange).toBeCalledTimes(0);

    const input = container.querySelector('input');
    expect(input).toBeInTheDocument();

    if (input) {
      fireEvent.change(input, { target: { value: 'Changed' } });
      expect(onChange).toBeCalledTimes(1);
      expect(onChange).toBeCalledWith('Changed');
    }
  });

  it('should handle textarea change', () => {
    const onChange = jest.fn();

    const { container } = render(
      withMockedProviders(
        <FormField
          label="Label"
          field={{
            type: 'text',
            value: 'Content',
            textarea: true,
            onChange,
          }}
        />,
      ),
    );
    expect(onChange).toBeCalledTimes(0);

    const textarea = container.querySelector('textarea');
    expect(textarea).toBeInTheDocument();

    if (textarea) {
      fireEvent.change(textarea, { target: { value: 'Changed' } });
      expect(onChange).toBeCalledTimes(1);
      expect(onChange).toBeCalledWith('Changed');
    }
  });

  it('should handle dropdown change', () => {
    const onChange = jest.fn();

    const mockedDropdown = jest.spyOn(Dropdown, 'Dropdown');
    expect(mockedDropdown).toBeCalledTimes(0);

    render(
      withMockedProviders(
        <FormField
          label="Label"
          field={{
            type: 'dropdown',
            value: '1',
            options: [
              { id: '1', title: 'Item 1', value: '1' },
              { id: '2', title: 'Item 2', value: '2' },
              { id: '3', title: 'Item 3', value: '3' },
            ],
            onChange,
          }}
        />,
      ),
    );
    expect(onChange).toBeCalledTimes(0);
    expect(mockedDropdown).toBeCalledTimes(1);

    // trigger Dropdown onChange
    mockedDropdown.mock.calls[0][0].onSelect('Changed');

    expect(onChange).toBeCalledTimes(1);
    expect(onChange).toBeCalledWith('Changed');
  });
});
