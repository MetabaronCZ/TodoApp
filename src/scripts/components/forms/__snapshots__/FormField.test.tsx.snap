// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/forms/FormField should render correctly 1`] = `
.c2 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  color: #222;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c5 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: block;
  width: 100%;
  padding: 5px 10px;
  background: #fff;
  border: none;
  outline: none;
}

.c5:hover:not(:disabled) {
  background-color: #fcfcfc;
}

.c5:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c5:disabled {
  color: #c0c0c0;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  gap: 10px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100px;
  min-height: 30px;
}

.c3 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <label
        class="c2"
        for="form-field-0"
      >
        Label:
      </label>
    </div>
    <div
      class="c3"
    >
      <div
        class="c4"
      >
        <input
          class="c5"
          id="form-field-0"
          type="text"
          value="Content"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`components/forms/FormField should render dropdown 1`] = `
.c2 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  color: #222;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  gap: 10px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100px;
  min-height: 30px;
}

.c3 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <label
        class="c2"
        for="form-field-3"
      >
        Label:
      </label>
    </div>
    <div
      class="c3"
    />
  </div>
</div>
`;

exports[`components/forms/FormField should render text input 1`] = `
.c2 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  color: #222;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c5 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: block;
  width: 100%;
  padding: 5px 10px;
  background: #fff;
  border: none;
  outline: none;
}

.c5:hover:not(:disabled) {
  background-color: #fcfcfc;
}

.c5:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c5:disabled {
  color: #c0c0c0;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  gap: 10px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100px;
  min-height: 30px;
}

.c3 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <label
        class="c2"
        for="form-field-1"
      >
        Label:
      </label>
    </div>
    <div
      class="c3"
    >
      <div
        class="c4"
      >
        <input
          class="c5"
          id="form-field-1"
          type="text"
          value="Content"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`components/forms/FormField should render textarea 1`] = `
.c2 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  color: #222;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c5 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: block;
  width: 100%;
  padding: 5px 10px;
  background: #fff;
  border: none;
  outline: none;
  min-height: 110px;
  resize: vertical;
}

.c5:hover:not(:disabled) {
  background-color: #fcfcfc;
}

.c5:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c5:disabled {
  color: #c0c0c0;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  gap: 10px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100px;
  min-height: 30px;
}

.c3 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <label
        class="c2"
        for="form-field-2"
      >
        Label:
      </label>
    </div>
    <div
      class="c3"
    >
      <div
        class="c4"
      >
        <textarea
          class="c5"
          id="form-field-2"
        >
          Content
        </textarea>
      </div>
    </div>
  </div>
</div>
`;
